src/charm.py

```diff
@@ -131,6 +132,10 @@ def __init__(self, *args):
             endpoint_path=f"{external_url.path}/api/v1/write",
         )
 
+        self._remote_write_consumer = PrometheusRemoteWriteConsumer(
+            self, extra_alerts_callable=self.metrics_consumer.alerts
+        )
```

Comment:
I'm a bit concerned about behaviour in unintended scenarios:
- What happens to the `self.remote_write_provider` of a middle-prom in a tiered scenario? Should we have a config option to switch between modes?
- Do we need to include here alerts coming from `self.remote_write_provider`?

Reply:
Is having a middle prometheus in a tiered scenario even a valid use case right now? Grafana Agent sort of does this (scrape one and push to the other), remote write provider alerts are already handled via `on.receive_remote_write`, too.

In particular, the "tiered prom" use case should not be recommended. Grafana Agent does this, or remote writing with Mimir eventually. Bad Things ™️ happen to Prometheus once it starts remote writing instead of something else scraping it (via `/federate` or other). Primarily that it starts sort-of discarding alerts. Grafana Agent sidesteps that.

Reply:
By "middle prom" I was referring to either COS1::Prometheus or COS2::Prometheus in the deployment scenario:
What happens if, in addition to alertmanager, someone relates to prom over remote-write?
In the POV of the local COS cluster, they probably expect their prom to be able to intake remote write.

```mermaid
graph LR
subgraph COS1
  am1[Alertmanager]
  prom1[Prometheus]
  am1 --- prom1
end
subgraph COS2
  am2[Alertmanager]
  prom2[Prometheus]
  am2 --- prom2
end
subgraph COS[Main COS]
  am[Alertmanager]
  prom[Prometheus]
  prom --- am
end
pd1[PagerDuty] --- am1
pd2[PagerDuty] --- am2
am --- pd[PagerDuty]
prom1 --- |remote write, rules| prom
prom2 --- |remote write, rules| prom
```

Reply:
do you mean like this?
I am a bit lost :)

```mermaid
graph LR
subgraph COS1
  am1[Alertmanager]
  prom1[Prometheus]
  am1 --- prom1
end
subgraph COS1.2
  am1.2[Alertmanager]
  prom1.2[Prometheus]
  am1.2 --- prom1.2
end
subgraph COS2
  am2[Alertmanager]
  prom2[Prometheus]
  am2 --- prom2
end
subgraph COS[Main COS]
  am[Alertmanager]
  prom[Prometheus]
  prom --- am
end
pd1[PagerDuty] --- am1
pd2[PagerDuty] --- am2
am --- pd[PagerDuty]
prom1 --- |remote write, rules| am1.2
prom1.2 --- |remote write, rules| prom
prom2 --- |remote write, rules| prom
```

Reply:
Like this:

```mermaid
graph LR
subgraph COS1
  am1[Alertmanager]
  prom1[Prometheus]
  am1 --- prom1
end

subgraph COS[Main COS]
  am[Alertmanager]
  prom[Prometheus]
  prom --- am
end
pd1[PagerDuty] --- am1
app1 ---|scrape| prom1
app2 ---|remote-write| prom1

am --- pd[PagerDuty]
prom1 --- |remote write, rules| prom
```

Reply:
No, this is exactly the case we do not want. Grafana Agent does this in a much better way. Remote write does not function this way without something like this patch. Prometheus evaluates alert rules itself, and speaks directly to Alertmanager. The entire experience is unintuitive and terrible. 

The default expectation from Prometheus is that, in your scenario, COS1 Prometheus will evaluate all rules it knows about, write the evaluation status to disk (which is not part of the WAL AFAIK and won't be sent across remote write -- it'll directly trigger any Alertmanagers it knows about). 

As a "general" expectation, the "Main COS" in a remote-write scenario would not receive any alert rules at all, and would need "fresh" ones (or recording rules could be sent across remote write, because those are in the WAL; or if it's Thanos/Cortex/Mimir, it can use a Ruler to ask Prometheus what's firing and go from there).

As a "general" expectation, the Prometheus sending remote write would not be connected to any Alertmanagers at all (because those Alertmanagers are isolated and can't dedupe, so whomever gets paged will get the same thing multiple times). It's handling evaluation only, and the "Main COS" would do all of the "actual" talking to whatever API the alerts are sent from.

I mentioned that Grafana Agent (or `/federate` in the "old days") is a common intermediary because it can relabel to adjust the granularity and mitigate this somewhat (so at least the evaluation may be different). If we're going with a design such as "alert rules are forwarded when Prom is related over remote_write", adding toggles to complicate this relation further is very strictly worse and higher complexity than using a small integrator charm for "remote write with alerts" or "remote write without alerts", depending on preference, or doing it for Grafana agent, or... whatever.